File: src/main/java/com/skysoft/mixin/LivingEntitySwingMixin.java
package com.skysoft.mixin;

import com.skysoft.features.helditem.HeldItemSwing;
import net.minecraft.world.entity.LivingEntity;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

@Mixin(LivingEntity.class)
public class LivingEntitySwingMixin {
    @Inject(method = "getCurrentSwingDuration", at = @At("RETURN"), cancellable = true)
    private void skysoft$modifyHeldItemSwingDuration(CallbackInfoReturnable<Integer> cir) {
        int duration = HeldItemSwing.duration((LivingEntity) (Object) this, cir.getReturnValue());
        cir.setReturnValue(duration);
    }
}

File: src/main/kotlin/com/skysoft/mixin/LivingEntitySwingMixin.kt
package com.skysoft.mixin

import com.skysoft.features.helditem.HeldItemSwing
import net.minecraft.world.entity.LivingEntity
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

@Mixin(LivingEntity::class)
open class LivingEntitySwingMixin {
    @Inject(method = ["getCurrentSwingDuration"], at = [At("RETURN")], cancellable = true)
    protected fun skysoftModifyHeldItemSwingDuration(cir: CallbackInfoReturnable<Int>) {
        val duration = HeldItemSwing.duration(this as LivingEntity, cir.returnValue)
        cir.setReturnValue(duration)
    }
}
